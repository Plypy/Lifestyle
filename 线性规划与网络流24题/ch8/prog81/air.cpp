#include <iostream>
#include <fstream>
#include <cstdlib>
#include <cstring>
using namespace std;
ifstream fin("air.in");
ofstream fout("air.out");
const int MAXN = 110;
bool g[MAXN][MAXN];
bool used[MAXN];
int pal[MAXN];
int n, m;

void load() {
    fin >> m >> n;
    n -= m;
    int a, b;
    while (true) {
        fin >> a >> b;
        if (-1 == a)
            break;
        b -= m;
        g[a][b] = true;
    }
}
int tot;
bool suc[MAXN];
bool find(int u) {
    for (int v = 1; v <= n; ++v)
        if (g[u][v] && !used[v]) {
            used[v] = true;
            if (!pal[v] || find(pal[v])) {
                pal[v] = u;
                return true;
            }
        }
    return false;
}

void work() {
    for (int i = 1; i <= m; ++i) {
        memset(used,0,sizeof(used));
        if (find(i)) {
            ++tot;
            suc[i] = true;
        }
    }
}

void output() {
    if (tot)
        fout << tot << endl;
    else
        fout << "No Solution!" << endl;
    for (int i = 1; i <= m; ++i)
        if (suc[i])
            fout << i << ' ' <<  n+pal[i] << endl;
}

int main() {
    load();
    work();
    output();
    return 0;
}
