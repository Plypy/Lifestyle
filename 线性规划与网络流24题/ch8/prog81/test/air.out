24
1 45
2 47
3 44
4 30
5 26
6 27
7 29
8 38
9 33
10 48
11 34
12 31
13 42
14 28
15 43
16 25
17 32
18 41
19 46
20 37
21 40
22 39
23 35
24 36
